### PR TITLE
Optimize large-file download path by reusing presigned URLs

### PR DESCRIPTION
**Pull request by JaySon-Huang**
## Summary

The main change is to stop resolving a fresh redirect for every download range on large-file reads. Instead, the client resolves the presigned object URL once and reuses that URL for all parallel `Range GET` requests.

In addition to that main path change, this branch also includes the adjacent changes that are part of the current branch diff:

- reuse multipart upload buffers to reduce upload-path allocation churn
- emit benchmark download summaries through the CLI logger instead of a separate benchmark-only output path
- centralize CLI log path helpers and write logs under `.drive9/cli/drive9-cli.log`
- expose the build git hash in `drive9 version` so bench binaries can be verified explicitly

## Design Basis

The implemented shape matches the design intent:

- keep server protocol unchanged
- resolve the presigned download target once in `pkg/client`
- reuse that object URL across the parallel downloader worker pool
- keep small-file behavior unchanged
- keep coarse download summary available for benchmark collection

### 1. Reuse a single presigned URL for parallel large-file downloads

`pkg/client/transfer.go`

- adds a dedicated large-file download path that resolves the read target once
- issues strict parallel `Range GET` requests directly against the object URL
- keeps the single-stream path for small files
- records coarse download summary metadata for benchmark analysis

This is the core change for the `reuse-presigned-download-url` proposal.

### 2. Wire download summary through the CLI log stream

`cmd/drive9/cli/cp.go`
`pkg/logger/factory.go`

- `drive9 fs cp` downloads now emit `download_summary` through the normal CLI logger when `DRIVE9_CLI_LOG_ENABLED=true`
- benchmark parsing now relies on the CLI log file rather than a separate benchmark-only stderr contract
- the CLI logger now owns a single canonical path: `.drive9/cli/drive9-cli.log`
- the logger creates the log directory automatically before opening the file

### 3. Reuse multipart upload buffers

`pkg/client/transfer.go`

- reuses one part-sized buffer per in-flight worker on multipart upload paths
- avoids allocating a fresh part buffer on every iteration
- keeps concurrency bounded by the existing memory-based policy

### 4. Expose build commit hash in `drive9 version`

`cmd/drive9/main.go`
`Makefile`

- injects the git hash at build time
- prints the hash in `drive9 version`
- makes it practical to verify that the binary on the bench host matches the expected source commit

## Why This Change

The previous parallel-download experiment proved that simply turning on parallel ranges was not enough. It regressed because each range still went through:

1. `GET /v1/fs/<path>`
2. redirect resolution
3. object-store `Range GET`

For a 1 GiB file split into 128 ranges, that means repeating the control-plane redirect path 128 times.

This PR fixes that shape directly by resolving the presigned object URL once and reusing it for the entire parallel download.

## Benchmark

PR comparison benchmark was rerun on the bench instance after cleaning the earlier invalid runs.

Baseline commit:

- `e1e57ee3136a2d17b948a492f4b1e904ce45bae4`

Benchmarked optimized commit:

- `da2501915f317688c4c089d4c939f04ec375342c`

Note: the branch head also contains `66552b63fc92d0760ceb939ecad1acce2f44a2f6`, which only adds build-hash reporting in `drive9 version`. The performance rerun was taken on `da25019`, which is the last behavior-relevant commit for the download/logging path.

### Drive9

| Metric | Baseline | After | Delta |
| --- | ---: | ---: | ---: |
| Upload elapsed | 8.245s | 8.131s | -1.38% |
| Upload throughput | 124.221 MiB/s | 125.945 MiB/s | +1.39% |
| Download elapsed | 14.421s | 3.474s | -75.91% |
| Download throughput | 71.018 MiB/s | 294.779 MiB/s | +315.08% |
| End-to-end elapsed | 22.666s | 11.605s | -48.80% |
| End-to-end throughput | 45.178 MiB/s | 88.237 MiB/s | +95.31% |

### S3 reference

| Metric | Baseline | After | Delta |
| --- | ---: | ---: | ---: |
| Upload elapsed | 2.206s | 2.212s | +0.27% |
| Upload throughput | 464.637 MiB/s | 463.256 MiB/s | -0.30% |
| Download elapsed | 2.548s | 2.524s | -0.94% |
| Download throughput | 401.966 MiB/s | 405.859 MiB/s | +0.97% |
| End-to-end elapsed | 4.753s | 4.736s | -0.36% |
| End-to-end throughput | 215.463 MiB/s | 216.221 MiB/s | +0.35% |

### Drive9 vs S3 ratio

| Metric | Baseline | After |
| --- | ---: | ---: |
| Upload | 0.268 | 0.272 |
| Download | 0.177 | 0.727 |
| End-to-end | 0.210 | 0.408 |

Interpretation:

- the gain is concentrated on the download path, which is exactly what this proposal targets
- upload remains effectively flat
- the S3 reference remains stable, which makes the before/after comparison credible

## Scope Notes

Although the main topic of this branch is the presigned-URL download-path fix, the branch currently also includes adjacent upload-path and observability/build-traceability changes. This description reflects the actual branch contents rather than only the original proposal title.

## Summary by CodeRabbit

* **New Features**
  * Parallel large-file downloads with performance tracking
  * Download summary events emitted to CLI logs
  * CLI shows Git commit hash in version output
  * CLI logging now exposes helpers to determine log path and enabled state (creates log file/dir)

* **Tests**
  * Added tests for download behaviors, summary logging, and version string

* **Documentation**
  * Clarified that enabling CLI structured logs includes transfer summary events

### CODE REVIEW

**Comment by qiffang** [comment]
## Code Review Summary

Reviewed by: **openclaw-work**, **codex**

### 📋 Overall Assessment: **LGTM with minor suggestions**

Excellent performance optimization! The benchmark shows **+315% download throughput** improvement. The core design is sound and implementation is clean.

---

## 🎯 Key Changes Reviewed

### 1. Presigned URL Reuse ✅

**Core optimization**: Resolve redirect once, reuse for all parallel Range GETs.

```go
target, err := c.resolveReadTarget(ctx, remotePath)  // One resolve
for task := range tasks {
    c.readObjectRangeStrict(ctx, target.objectURL, ...)  // Reuse URL
}
```

**Strengths**:
- Server protocol unchanged (pure client-side optimization)
- `readObjectRangeStrict` properly validates HTTP 206 vs 200
- Detects when Range header is ignored (returns error instead of silent corruption)

### 2. Upload Buffer Pool ✅

- Pre-allocates buffers to avoid per-part allocation churn
- `put()` restores original length to prevent slice corruption on reuse
- Good test coverage for edge cases

### 3. CLI Log Path Migration ✅

- Old: `~/.dat/log/dat9-cli.log`
- New: `~/.drive9/cli/drive9-cli.log`

Part of the dat9 → drive9 rename. Consider noting this in release documentation.

### 4. Build Git Hash ✅

`drive9 version` now shows commit hash — useful for debugging and benchmark verification.

---

## ⚠️ Suggestions

### 1. Presigned URL Expiration (from codex + openclaw-work)

**Risk**: S3 presigned URLs typically expire in 5-15 minutes. For very large files (100GB+) or slow networks, later Range requests may hit 403/expired.

**Recommendations**:
- Option A: Carry expiration time in `readTarget`, proactively refresh before expiry
- Option B: On 403/SignatureExpired, retry once with fresh `resolveReadTarget()`
- Option C: Document the limitation for now, address in follow-up PR

### 2. Cross-Range Consistency (from codex)

**Concern**: If the object changes mid-download, different ranges may return inconsistent data.

**Recommendation**: Consider capturing ETag/VersionId during `resolveReadTarget()` and adding `If-Match` header to Range requests for strict consistency (optional enhancement).

### 3. Buffer Pool Guard (from codex)

The `put()` method assumes callers only return pool-owned buffers. Consider adding a capacity check guard to prevent panics if misused in the future.

---

## ✅ Test Coverage

New tests cover core scenarios:
- `TestDownloadToFileWithSummaryReusesPresignedURL` — verifies single redirect resolution
- `TestDownloadToFileWithSummaryFailsWhenRangeNotHonored` — verifies strict Range validation
- `TestUploadBufferPoolRestoresFullLengthOnPut` — verifies buffer reuse safety

---

## 📊 Benchmark Verification

The PR includes detailed before/after benchmarks:
- Download throughput: 71 → 295 MiB/s (**+315%**)
- Download elapsed: 14.4s → 3.5s (**-76%**)
- Upload unchanged (as expected)
- S3 reference stable (validates measurement credibility)

---

## 🎉 Conclusion

**Approved with suggestions**. This is a high-quality optimization with significant impact. The suggestions above are enhancements for edge cases, not blockers.

Great work @JaySon-Huang! 🚀

**Comment by JaySon-Huang** [comment]
Follow-up on the review suggestions:

- opened #138 for presigned URL expiry handling during parallel downloads
- opened #139 for cross-range consistency / object identity enforcement
- added code comments in `pkg/client/transfer.go` to make the current boundary explicit in this PR

The current implementation now documents that:

- one resolved presigned URL is assumed to stay valid for the lifetime of a single download
- cross-range consistency is best-effort today because we do not yet bind range requests to an `ETag` / `VersionId`

I kept both as follow-up issues instead of expanding PR #137 scope, since they need separate behavior and failure-semantics design.

**Comment by JaySon-Huang** [comment]
Addressed the `uploadBufferPool.put()` guard suggestion in this PR.

Changes in `b020637`:

- added an ownership note on `uploadBufferPool.put()` to make the assumption explicit: the buffer is expected to come from the same pool via `get()`
- added a small guard so a foreign buffer with insufficient capacity is dropped instead of panicking on `buf[:p.size]`
- added `TestUploadBufferPoolPutDropsForeignShortBuffer` to cover that misuse case

Validation:

- `make test TEST_P=1` passes locally

I kept this as a small local hardening change rather than expanding the broader download-path scope.

**Comment by JaySon-Huang** [comment]
@coderabbitai review

File: cmd/drive9/cli/cp.go [hunk 1]
```diff
@@ -134,3 +133,27 @@ func printProgress(partNumber, totalParts int, bytesUploaded int64) {
...
+		zap.String("mode", summary.Mode),
+		zap.Int("concurrency", summary.Concurrency),
+		zap.Int64("chunk_size_bytes", summary.ChunkSizeBytes),
+		zap.Int("range_count", summary.RangeCount),
+		zap.Time("started_at", summary.StartedAt),
+		zap.Time("finished_at", summary.FinishedAt),
+		zap.Float64("elapsed_seconds", summary.ElapsedSeconds),
+		zap.String("remote_path", summary.RemotePath),
+		zap.String("local_path", summary.LocalPath),
+	)
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply cmd/drive9/cli/cp.go:158]: Verified this path does not panic with the current logger implementation, so we are not taking the suggested fix in this PR.

`pkg/logger/logger.go` installs a process-wide production logger in `init()`, and `logger.Info(...)` falls back to that global logger via `FromContext()` / `L()` even when `NewCLILogger()` fails. The practical risk here is that benchmark summary output can leak to stderr or the default log sink instead of the CLI log file, not that the download command crashes.

We are accepting that benchmark-only output-contract risk for now and keeping this PR focused on the large-file transfer path changes.

File: pkg/logger/factory.go [hunk 2]
```diff
@@ -15,16 +15,38 @@ func NewServerLogger() (*zap.Logger, error) {
...
-	logDir := filepath.Join(home, ".dat", "log")
 	if err := os.MkdirAll(logDir, 0o755); err != nil {
 		return nil, fmt.Errorf("create log dir: %w", err)
 	}
-	logPath := filepath.Join(logDir, "dat9-cli.log")
+	logPath, err := CLILogPath()
+	if err != nil {
+		return nil, err
+	}
 	f, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply pkg/logger/factory.go:50]: Keeping the existing CLI log file permissions in this PR.

The current benchmark environment does not require tightening these file modes, and we do not want to mix that operational policy change into this large-file transfer performance PR. Treating this as out of scope for now rather than a blocker for the download-path work.

[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply pkg/logger/factory.go:50]: @coderabbitai yes, you can open a GitHub issue to track this issue!

**Review by qiffang — APPROVED** [review]
